File: j.py
import aoc
from math import atan2, pi
class Vect(tuple): #to overload '-' operator
    def __sub__(self, other):
        return Vect(x-y for x,y in zip(self, other))

ma = 0
s = aoc.sreadlines("i.10")
riods = [Vect([x,y]) for y in range(len(s)) for x in range(len(s[0])) if s[y][x] == '#']
def angle(x,y):
    #note: the axis of y is flipped
    a = atan2(y,x)
    if x < 0 and y < 0: a += 2*pi
    return a
for cur in riods: #each calculation here is repeated unnecessarily once
    angs = set()
    for comp in riods:
        if comp == cur: continue
        ang = angle(*comp-cur)
        if ang not in angs: #this does not find the closest '#', but it does count correctly
            angs.add(ang)
    v = len(angs)
    if ma < v:
        ma = v
        best = cur
print ma

def sqsum(t): sum(map(lambda v: v*v, t))
targets = set(riods)
targets.remove(best)
i = 0
while 1:
    angs = {}
    for comp in targets:
        ang = angle(*comp-best)
        if ang not in angs or sqsum(angs[ang]-best) > sqsum(comp-best):
            angs[ang] = comp
    v = len(angs)
    
    if i+v < 200:
        for k in angs: targets.remove(k)
        i += v
    else:
        ls = sorted(angs.items())
        _, ans = ls[199-i]
        print ans[0]*100+ans[1]
        break

File: j.js
const aoc = require('./aoc');
const s = aoc.sreadlines("i.10");
function sub(p, t) {
  return { x: p.x-t.x, y: p.y-t.y };
}
let riods = [];
s.forEach((l,y) => riods = riods.concat(
  Array.from(l, (c,x) => c == '#' ? {x: x, y: y} : null).filter(v=>v)));
function angle(p) {
    const y = p.y, x = p.x;
    let a = Math.atan2(y,x);
    if (x < 0 && y < 0) a += 2*Math.PI;
    return a;
}
function Between(f, p, t) {
    const diff = sub(p,t);
    return f(diff);
}
let ma = 0, best;
riods.forEach(cur => {
  const angs = new Set();
  riods.forEach(comp => {
    if (comp === cur) return;
    const ang = Between(angle, comp, cur);
    if (!angs.has(ang)) angs.add(ang);
  });
  const v = angs.size;
  if (ma < v) {
    ma = v;
    best = cur;
  }
});
console.log(ma);
function sqsum(p) {
  return Object.values(p).map(v => v*v).reduce(aoc.sum);
}
const targets = new Set(riods);
targets.delete(best);
let angs = {}, i = 0;
for (let v = 0; i+v < 200;) {
  i += v;
  for (const k in angs) targets.delete(Number(k));
  for (const comp of targets) {
    const ang = Between(angle, comp, best);
    if (!angs.hasOwnProperty(ang) || Between(sqsum, angs[ang], best) > Between(sqsum, comp, best)) angs[ang] = comp;
  }
  v = angs.size;
}
let ls = Object.entries(angs);
ls.sort((a,b) => a[0] < b[0] ? -1 : 1);
const ans = ls[199-i][1];
console.log(ans.y+100*ans.x)
